deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/AsyncDataSetIterator.java

```diff
@@ -113,291 +93,282 @@ public AsyncDataSetIterator(DataSetIterator iterator, int queueSize, BlockingQue
+     * iterator that does asynchronous prefetching. For example, it would not make sense to use asynchronous
+     * prefetching for the following types of iterators:
+     * (a) Iterators that store their full contents in memory already
+     * (b) Iterators that re-use features/labels arrays (as future next() calls will overwrite past contents)
+     * (c) Iterators that already implement some level of asynchronous prefetching
+     * (d) Iterators that may return different data depending on when the next() method is called
+     *
+     * @return true if asynchronous prefetching from this iterator is OK; false if asynchronous prefetching should not
+     * be used with this iterator
+     */
     @Override
     public boolean asyncSupported() {
         return false;
     }
 
+    /**
+     * Resets the iterator back to the beginning
+     */
     @Override
-    public synchronized void reset() {
-        if (!resetSupported())
-            throw new UnsupportedOperationException(
-                    "Cannot reset Async iterator wrapping iterator that does not support reset");
-        //Complication here: runnable could be blocking on either baseIterator.next() or blockingQueue.put()
-        runnable.killRunnable = true;
-        if (runnable.isAlive.get()) {
-            thread.interrupt();
-        }
-        //Wait for runnable to exit, but should only have to wait very short period of time
-        //This probably isn't necessary, but is included as a safeguard against race conditions
-        try {
-            runnable.runCompletedSemaphore.tryAcquire(5, TimeUnit.SECONDS);
-        } catch (InterruptedException e) {
-        }
-
-        //Clear the queue, reset the base iterator, set up a new thread
-        blockingQueue.clear();
-        baseIterator.reset();
-        runnable = new IteratorRunnable(baseIterator.hasNext(), this.workspace);
-        thread = runnable;
+    public void reset() {
+        shutdown();
+        backedIterator.reset();
+        shouldWork.set(true);
+        this.thread = new AsyncPrefetchThread(buffer, backedIterator, terminator, workspace);
 
+        /**
+         * We want to ensure, that background thread will have the same thread->device affinity, as master thread
+         */
         Integer deviceId = Nd4j.getAffinityManager().getDeviceForCurrentThread();
         Nd4j.getAffinityManager().attachThreadToDevice(thread, deviceId);
 
         thread.setDaemon(true);
         thread.start();
+
+        nextElement = null;
+    }
+
+
+    public void shutdown(){
+        buffer.clear();
+
+
+        if (thread != null)
+            thread.interrupt();
+        try {
+            // Shutdown() should be a synchronous operation since the iterator is reset after shutdown() is
+            // called in AsyncLabelAwareIterator.reset().
+            if (thread != null)
+                thread.join();
+        } catch (InterruptedException e) {
```

Comment:
This usually shouldn't happen, but should we be silently swallowing the exception here if it does?